src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -1504,34 +1521,44 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: SlickDataSo
     }
   }
 
-  private def withSubmissionEntities(submissionRequest: SubmissionRequest, workspaceContext: SlickWorkspaceContext, rootEntityType: String, dataAccess: DataAccess)(op: (Seq[Entity]) => ReadWriteAction[PerRequestMessage]): ReadWriteAction[PerRequestMessage] = {
+  private def withSingleEntityRec(entityType: String, entityName: String, workspaceContext: SlickWorkspaceContext, dataAccess: DataAccess)(op: (Seq[EntityRecord]) => ReadWriteAction[PerRequestMessage]): ReadWriteAction[PerRequestMessage] = {
+    val entityRec = dataAccess.entityQuery.findEntityByName(workspaceContext.workspaceId, entityType, entityName).result
+    entityRec flatMap { entities =>
+      if (entities.isEmpty) {
+        DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.NotFound, s"No entity of type ${entityType} named ${entityName} exists in this workspace.")))
+      } else if (entities.size == 1) {
+        op(entities)
+      } else {
+        DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.NotFound, s"More than one entity of type ${entityType} named ${entityName} exists in this workspace?!")))
+      }
+    }
+  }
+
+  def withSubmissionEntityRecs(submissionRequest: SubmissionRequest, workspaceContext: SlickWorkspaceContext, rootEntityType: String, dataAccess: DataAccess)(op: (Seq[EntityRecord]) => ReadWriteAction[PerRequestMessage]): ReadWriteAction[PerRequestMessage] = {
```

Comment:
Q: Should this (and `withFoo` functions in general) take a parameterised type and specify `op` as `whatever => T` and the return type as `T` (or perhaps `ReadWriteAction[T]`), rather than specifying it has to be a `ReadWriteAction[PerRequestMessage]` specifically? This bit me [in tests, where I had to create a `RequestComplete` I didn't actually need or use so the typechecker would be happy](https://github.com/broadinstitute/rawls/pull/401/files#diff-412c7e893a8c21e01efe9a7c93578323R190).